In [29]:
import pandas as pd
import numpy as np
df = pd.read_csv("../data/processed/elevator_clean.csv")

In [30]:
df.to_csv('../data/processed/elevator_clean.csv', index=False)

In [31]:
features = [
    "revolutions",
    "humidity",
    "x1",
    "x2",
    "x3",
    "x4",
    "x5"
]

X = df[features]
y = df["vibration"]

features = [
    "revolutions",
    "humidity",
    "x1",
    "x2",
    "x3",
    "x4",
    "x5"
]

X = df[features]
y = df["vibration"]

In [32]:
split = int(len(df) * 0.80)

y_train = y.iloc[:split]
y_test = y.iloc[split:] # train-test split

In [33]:
# Calculate average vibration from training data
mean_value = y_train.mean()

# Predict the same value for every test sample
mean_predictions = np.full(len(y_test), mean_value) # it creates an array of the same length as y_test, filled with the mean_value

# Create some fake good predictions only for comparison
good_predictions = y_test + np.random.normal(0, 5, len(y_test)) # in simple terms, it adds some random noise to the actual test values to simulate a model that predicts well
print(mean_predictions)
print(good_predictions)

[29.29064777 29.29064777 29.29064777 ... 29.29064777 29.29064777
 29.29064777]
89600     -1.994578
89601     34.842433
89602      0.750685
89603     41.986032
89604      0.333892
            ...    
111996     7.250063
111997    19.449576
111998    19.709007
111999    21.648815
112000     4.875139
Name: vibration, Length: 22401, dtype: float64


In [34]:
# then we will calculate MAE, RMSE, and R2 score for both mean_predictions and good_predictions to compare their performance.
errors = abs(y_test - mean_predictions)
mae = np.mean(errors)
print(f'Mean Absolute Error (MAE) for mean predictions: {mae}')

Mean Absolute Error (MAE) for mean predictions: 11.172826601112355


In [35]:
# Calculate RMSE for mean predictions
squared_errors = (y_test - mean_predictions) ** 2
rmse = np.sqrt(np.mean(squared_errors))
print(f'Root Mean Squared Error (RMSE) for mean predictions: {rmse}')

Root Mean Squared Error (RMSE) for mean predictions: 14.985809101903333


In [36]:
# R^2 score for mean predictions
ss_total = np.sum((y_test - np.mean(y_test)) ** 2)
ss_residual = np.sum((y_test - mean_predictions) ** 2)
r2 = 1 - (ss_residual / ss_total)
print(f'R^2 Score for mean predictions: {r2}')

R^2 Score for mean predictions: -0.05147880026285767


### Baseline Mean Prediction is
+ MAE is 11.17
+ RMSE is 14.98
+ R2 score is -0.051


### Linear Regression
+ Mean Absolute Error: 10.9689
+ Root Mean Squared Error: 16.2670
+ R^2 Score: -0.2390

### Random Forest with Lag Features
+ MAE: 3.4104
+ MSE: 86.8938
+ RMSE: 9.3217
+ R^2 Score: 0.5931

### Random Forest with Regmime Split
+ MAE: 4.9501
+ MSE: 104.5682
+ RMSE: 10.2259
+ R2: 0.5103

### Random Forest with Rolling Stats
+ MAE : 2.9792255595855504
+ RMSE : 7.378123483933775
+ R2 : 0.7449624965478825

### XGBoost
+  MAE  = 2.9375
+ RMSE = 7.2720
+ R²   = +0.7522